plugins/inputs/vsphere/vsan.go

```diff
@@ -0,0 +1,539 @@
+			vals := strings.Split(em.EntityRefId, ":")
+			var entityName, uuid string
+			if len(vals) == 1 {
+				entityName, uuid = vals[0], ""
+			} else {
+				entityName, uuid = vals[0], vals[1]
+			}
+
+			buckets := make(map[string]metricEntry)
+			tags := populateCMMDSTags(tags, entityName, uuid, cmmds)
+			var timeStamps []time.Time
+			// 1. Construct a timestamp list from sample info
+			formattedEntityName := hyphenReplacer.Replace(entityName)
+			for _, t := range strings.Split(em.SampleInfo, ",") {
+				tsParts := strings.Split(t, " ")
+				if len(tsParts) >= 2 {
+					// The return time string is in UTC time
+					utcTimeStamp := fmt.Sprintf("%sT%sZ", tsParts[0], tsParts[1])
+					ts, ok := time.Parse(time.RFC3339, utcTimeStamp)
+					if ok != nil {
+						e.Parent.Log.Errorf("[vSAN] Failed to parse a timestamp: %s. Skipping", utcTimeStamp)
+						timeStamps = append(timeStamps, time.Time{})
+						continue
+					}
+					timeStamps = append(timeStamps, ts)
+				}
+			}
+			// 2. Iterate on each measurement
+			for _, counter := range em.Value {
+				metricLabel := internal.SnakeCase(counter.MetricId.Label)
+				// 3. Iterate on each data point.
+				for i, values := range strings.Split(counter.Values, ",") {
+					ts := timeStamps[i]
+					if ts == (time.Time{}) {
+						continue
+					}
+					// Organize the metrics into a bucket per measurement.
+					bKey := em.EntityRefId + " " + strconv.FormatInt(ts.UnixNano(), 10)
+					bucket, found := buckets[bKey]
+					if !found {
+						mn := vsanPerfMetricsName + e.Parent.Separator + formattedEntityName
+						bucket = metricEntry{name: mn, ts: ts, fields: make(map[string]interface{}), tags: tags}
+						buckets[bKey] = bucket
+					}
+					if v, err := strconv.ParseFloat(values, 32); err == nil {
+						bucket.fields[metricLabel] = v
+					}
+				}
+			}
+			if len(timeStamps) > 0 {
+				lastSample := timeStamps[len(timeStamps)-1]
+				if lastSample != (time.Time{}) && lastSample.After(latest) {
+					latest = lastSample
+				}
+			}
+			// We've iterated through all the metrics and collected buckets for each measurement name. Now emit them!
+			for _, bucket := range buckets {
+				acc.AddFields(bucket.name, bucket.fields, bucket.tags, bucket.ts)
+			}
+			count += len(buckets)
+		}
+	}
+	e.hwMarks.Put(hwMarksKeyPrefix+clusterRef.ref.Value, "generic", latest)
+	return nil
+}
+
+// queryDiskUsage adds 'FreeCapacityB' and 'TotalCapacityB' metrics to telegraf accumulator
+func (e *Endpoint) queryDiskUsage(ctx context.Context, vsanClient *soap.Client, clusterRef *objectRef, acc telegraf.Accumulator) error {
+	resp, err := vsanmethods.VsanQuerySpaceUsage(ctx, vsanClient,
+		&vsantypes.VsanQuerySpaceUsage{
+			This:    spaceManagerRef,
+			Cluster: clusterRef.ref,
+		})
+	if err != nil {
+		return err
+	}
+	fields := make(map[string]interface{})
+	fields["free_capacity_byte"] = resp.Returnval.FreeCapacityB
+	fields["total_capacity_byte"] = resp.Returnval.TotalCapacityB
```

Comment:
```suggestion
	fields := map[string]interface{}{
		"free_capacity_byte": resp.Returnval.FreeCapacityB,
		"total_capacity_byte": resp.Returnval.TotalCapacityB,
	}
```
Same for other static inits.